### PR TITLE
Add a query-builder to the /runs UI

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Also adds support for `to` and `from` (required for the default query on that tab).

For picking the date, leverages https://vaadin.com/directory/component/vaadinvaadin-date-picker

## Review Information
Visit /runs, try a few different queries.

### CODE REVIEW

**Comment by foolip** [comment]
Haven't looked at code, but I tried https://runs-builder-dot-wptdashboard-staging.appspot.com/runs a bit. I noticed that the URL isn't updated when submitting, which would be great for sharing links. Trivial to fix, I presume?

**Comment by foolip** [comment]
On the nitpicky side, I see that "EDIT" isn't vertically aligned in the box. Needs line-height?

**Comment by foolip** [comment]
An unrelated thing I noticed when looking for very old runs is that we don't show a year, so runs from 2017 might seem like they're from 2018 if one doesn't click through and check.

**Comment by lukebjerring** [comment]
Resolved the year by including it when it's not the current year.

**Comment by lukebjerring** [comment]
Regarding virtual nagivation: that's actually a little complicated and should be its own follow-up PR.

**Comment by lukebjerring** [comment]
EDIT button is vertically aligned for me (ff / chrome both). What browser? Screenshot?

**Comment by foolip** [comment]
I'm running Chrome on Mac. Screenshot:
screen shot 2018-11-16 at 5 59 27 pm

**Comment by foolip** [comment]
> Resolved the year by including it when it's not the current year.

Looks like this also causes the date to not be shown? Did it take too much space? (Fine if that's the reason and not accidental.)

**Comment by lukebjerring** [comment]
Try a cache-refresh?

https://runs-builder-dot-wptdashboard-staging.appspot.com/runs?from=2017-11-13T05%3A00%3A00.000Z&to=2018-01-13T05%3A00%3A00.000Z

**Comment by foolip** [comment]
https://runs-builder-dot-wptdashboard-staging.appspot.com/runs?from=2017-11-13T05%3A00%3A00.000Z&to=2018-01-13T05%3A00%3A00.000Z works for me, looks like year is shown the first time it changes, like days.

File: .travis.yml [hunk 1]
```diff
@@ -2,11 +2,6 @@
 # Use of this source code is governed by a BSD-style license that can be
 # found in the LICENSE file.
 
-notifications:
```
Comment (foolip) [inline .travis.yml]: Accidental change? Shouldn't someone get email for failures?

File: webapp/components/test-runs-query.html [hunk 2]
```diff
@@ -205,10 +211,17 @@
...
-            var diff = Math.floor((Date.now() - from) / 86400000);
-            return `Showing test runs from the last ${diff} days`;
+            const max = to || Date.now();
+            var diff = Math.floor((max - from) / 86400000);
+            const shortDate = date => date.toLocaleDateString(navigator.language, {
+              month: 'short',
+              day: 'numeric',
+            });
+            return to
+              ? `Showing test runs from ${shortDate(from)} to ${shortDate(to)}`
```
Comment (foolip) [inline webapp/components/test-runs-query.html]: Including the year if different or not the current year here would be good. I got "Showing test runs from Dec 15 to Nov 21" where the start was 2016 and end 2018. In that case I'd expect the year to be shown for both, so for simplicity maybe always show year for both dates or neither.

File: webapp/components/test-runs-query.html [hunk 3]
```diff
@@ -205,10 +211,17 @@
           this.setProperties(batchUpdate);
         }
 
-        computeResultsRangeMessage(from, maxCount) {
+        computeResultsRangeMessage(from, to, maxCount) {
           if (from) {
-            var diff = Math.floor((Date.now() - from) / 86400000);
-            return `Showing test runs from the last ${diff} days`;
+            const max = to || Date.now();
```
Comment (foolip) [inline webapp/components/test-runs-query.html]: `max` is only used to compute `diff` and diff is only used if `to` is falsy, so the purpose of this line is hard to parse. Seems like it could be replaced with `const max = Date.now()`. Maybe `if (to) { /* compute and return stuff */ } else { /* other stuff */ }` to make this bit more clear?

File: webapp/components/wpt-results.html [hunk 4]
```diff
@@ -628,6 +628,11 @@ <h3>
                     total: 0,
                   })),
                 };
+                row.diff = {
```
Comment (foolip) [inline webapp/components/wpt-results.html]: These and some following changes seem unrelated to this PR. Can you split out?

File: webapp/components/test/wpt-results.html [hunk 5]
```diff
@@ -64,82 +64,6 @@
             });
           });
         });
-
-        suite('computeDifferences(before, after)', () => {
```
Comment (foolip) [inline webapp/components/test/wpt-results.html]: Why is this bunch of tests deleted? This isn't the component this PR is changing.

File: .travis.yml (hunk 1 above)
Comment (lukebjerring) [reply .travis.yml]: Bad merge, maybe. Reverted.

File: webapp/components/test-runs-query.html (hunk 3 above)
Comment (lukebjerring) [reply webapp/components/test-runs-query.html]: Done.

File: webapp/components/test/wpt-results.html (hunk 5 above)
Comment (lukebjerring) [reply webapp/components/test/wpt-results.html]: Bad merge. reverted.

File: webapp/components/wpt-results.html (hunk 4 above)
Comment (lukebjerring) [reply webapp/components/wpt-results.html]: Bad merge. reverted.

**Comment by lukebjerring** [comment]
Something strange was going on so I reset to master and reverted the weirdness and forcepushed a squashed commit... PTAL

File: webapp/components/test-runs-query.html (hunk 2 above)
Comment (lukebjerring) [reply webapp/components/test-runs-query.html]: Done (when they differ)

**Comment by lukebjerring** [comment]
I can't repro the bug you describe. There's a chance that you're hitting public-cache issues, so make sure you give it a cache clear and see if it's still an issue.

**Comment by foolip** [comment]
Taking another look. Note that the alignment bug in https://github.com/web-platform-tests/wpt.fyi/pull/769#issuecomment-439458356 is still present.

File: webapp/components/test-runs-query.html [hunk 6]
```diff
@@ -205,10 +211,23 @@
...
+              };
+              if (from.getYear() !== to.getYear()) {
+                opts.year = 'numeric';
+              }
+              return date.toLocaleDateString(navigator.language, opts);
+            };
+            if (to) {
+              return `Showing test runs from ${shortDate(from)} to ${shortDate(to)}`;
+            }
+            const days = Math.floor((Date.now() - from) / 86400000);
```
Comment (foolip) [inline webapp/components/test-runs-query.html]: Thanks for changing this, much clearer!

**Comment by foolip** [comment]
I can still repro the bug with devtools open and "Disable cache" checked. Exact steps:
1. Visit https://runs-builder-dot-wptdashboard-staging.appspot.com/runs, see both Chrome Dev and Chrome stable logos, both Firefox Nightly and Firefox stable, and so on.
1. Click "EDIT", add "stable" to the labels field and click submit.
1. Now you see fewer runs, just stable browser logos.
1. Click "EDIT", note that the labels field is empty, change nothing, and submit
1. Now you see the original view.

**Comment by lukebjerring** [comment]
Bug should be fixed by #794
